rootfs/api/models/app.py

```diff
@@ -565,15 +554,38 @@ def deploy(self, release, force_deploy=False):
             self.log(err, logging.ERROR)
             raise ServiceUnavailable(err) from e
 
-        # Wait until application is available in the router
-        # Only run when there is no previous build / release
-        old = release.previous()
-        if old is None or old.build is None:
-            self.verify_application_health(**kwargs)
+        app_type = 'web' if 'web' in deploys else 'cmd' if 'cmd' in deploys else None
+        # Make sure the application is routable and uses the correct port Done after the fact to
+        # let initial deploy settle before routing traffic to the application
+        if deploys and app_type:
+            routable = deploys[app_type].get('routable')
+            port = deploys[app_type].get('envs', {}).get('PORT', None)
+            self._update_application_service(self.id, app_type, port, routable, service_annotations)  # noqa
+
+            # Wait until application is available in the router
+            # Only run when there is no previous build / release
+            old = release.previous()
+            if old is None or old.build is None:
+                self.verify_application_health(**deploys[app_type])
 
         # cleanup old release objects from kubernetes
         release.cleanup_old()
 
+    def _check_deployment_in_progress(self, deploys, force_deploy=False):
+        for scale_type, kwargs in deploys.items():
+            if force_deploy:
```

Comment:
If `force_deploy` isn't changed in this function, then it's a loop invariant and can be moved outside the loop. (Python's compiler isn't smart enough to do that for you.)
```python
if force_deploy:
    return

for scale_type, kwargs in deploys.items():
    # Is there an existing deployment in progress?
    name = self._get_job_id(scale_type)
...
```